apps/SageAccountWeb/src/components/useSourceApp.tsx

```diff
@@ -1,25 +1,135 @@
 import { sourceAppConfigTableID } from '../resources'
 import { SourceAppConfig } from './SourceAppConfigs'
-import { useSourceAppConfigs } from './useSourceAppConfigs'
+import {
+  STATIC_SOURCE_APP_CONFIG,
+  useSourceAppConfigs,
+} from './useSourceAppConfigs'
+import React, { useContext, useEffect, useMemo } from 'react'
+import { getSearchParam } from '../URLUtils'
+import { useLocalStorageValue } from '@react-hookz/web'
+import { SynapseTheme, useLastLoginInfoState } from 'synapse-react-client'
+import { createTheme, ThemeProvider } from '@mui/material/styles'
+import { sageAccountWebThemeOverrides } from '../style/theme'
+
+export type SourceAppContextType = SourceAppConfig
+
+const SOURCE_APP_ID_QUERY_PARAM_KEY = 'appId'
+const SOURCE_APP_ID_LOCALSTORAGE_KEY = 'sourceAppId'
+export const DEFAULT_SOURCE_APP_ID = 'SAGE'
+export const SYNAPSE_SOURCE_APP_ID = 'synapse.org'
+
+/**
+ * This must be exported to use the context in class components.
+ */
+export const SourceAppContext = React.createContext<SourceAppContextType>(
+  STATIC_SOURCE_APP_CONFIG,
+)
+
+export type SourceAppContextProviderProps = React.PropsWithChildren<{
+  sourceAppId?: string
+}>
+
+function useConfigureSourceAppFromQueryParams() {
+  const { value: localStorageAppId, set: setLocalStorageAppId } =
+    useLocalStorageValue(SOURCE_APP_ID_LOCALSTORAGE_KEY)
+
+  useEffect(() => {
+    const searchParamAppId = getSearchParam(SOURCE_APP_ID_QUERY_PARAM_KEY)
+    if (searchParamAppId) {
+      setLocalStorageAppId(searchParamAppId)
+    } else if (!localStorageAppId) {
+      // fallback to Sage Bionetworks
+      setLocalStorageAppId(DEFAULT_SOURCE_APP_ID)
+    }
+  }, [localStorageAppId, setLocalStorageAppId])
+}
+
+/**
+ * Synchronizes the current state of the `currentSourceAppName` localStorage value.
+ * The value may be used at a later date in the Login component to inform a user if they have a OneSage account
+ * that was used for a different application.
+ */
+export function useSynchronizeCurrentSourceAppNameForLogin(
+  sourceApp?: SourceAppConfig,
+) {
+  const { currentSourceAppNameState } = useLastLoginInfoState()
+  const { value: idFromLocalStorage } = useLocalStorageValue(
+    SOURCE_APP_ID_LOCALSTORAGE_KEY,
+  )
+
+  useEffect(() => {
+    if (idFromLocalStorage && sourceApp?.friendlyName) {
+      currentSourceAppNameState.set(sourceApp.friendlyName)
+    }
+  }, [idFromLocalStorage, currentSourceAppNameState, sourceApp?.friendlyName])
+}
+
+/**
+ * Provides context necessary for most components in SRC.
+ *
+ * The SynapseContextProvider must be wrapped in a react-query QueryClientProvider.
+ * @param props sourceAppId: An optional source app ID to use. If not provided, the source app ID will be read from localStorage, or default to 'SAGE'.
+ */
```

Comment:
Maybe a copy/paste issue -- should this comment be updated for `SourceAppProvider`? e.g. something like -- 

```suggestion
/**
 * Provides source app context necessary for most components in SageAccountWeb.
 *
 * The SourceAppProvider must be wrapped in a react-query QueryClientProvider.
 * @param props sourceAppId: An optional source app ID to use. If not provided, the source app ID will be read from localStorage, or default to 'SAGE'.
 */
```

Reply:
Yes, every time I create a new context I copy this boilerplate but forget to change the comments 😬

Thanks for catching